### PR TITLE
255 add mod solo

### LINKED ISSUE #255: Add mod Solo
Create a function in client which choose if the player want must play in solo or in multiplayer

Link serv lib to client and create new world which don't have all system link to the connection protocol (protocol 15)

And add a check to all systems that use server to not check a nonexistent Component or System

### PR DESCRIPTION
**Pull request by TomDUVAL-MAHE**
Add the possibility to play in solo or multiplayer mod

Have fun and comment all things that you think good !

Close #255

### CODE REVIEW

File: Client/ClientRoom.cpp [hunk 1]
```diff
@@ -113,11 +131,17 @@ ClientRoom::ClientRoom(std::string address, unsigned short port, std::string ser
     _password = "";
 }
 
-void ClientRoom::_initEcsGameData(void)
+void ClientRoom::_initEcsGameData(bool isSolo)
 {
-    _initSharedResources();
-    _initSystems();
-    _initEntities();
+    if (isSolo) {
```
Comment (31Nathan) [inline Client/ClientRoom.cpp:136]: Useless if

File: Client/ClientRoom.cpp [hunk 2]
```diff
@@ -217,6 +241,93 @@ void ClientRoom::_protocol15Answer(CommunicatorMessage connectionResponse)
...
+
+void ClientRoom::startSoloLoop()
+{
+    _initEcsGameData(true);
+    initSoloData();
+    _state = ClientState::IN_GAME;
+    std::signal(SIGINT, signalCallbackHandler);
+    while (_state != ClientState::ENDED && _state != ClientState::UNDEFINED) {
+        if (_state == ClientState::IN_GAME) {
+            _worldInstance.get()->runSystems(); /// WILL BE IMPROVED IN PART TWO (THREAD + CLOCK)
```
Comment (31Nathan) [inline Client/ClientRoom.cpp:265]: Comment Up to Date ?

File: Client/ClientRoom.cpp [hunk 3]
```diff
@@ -217,6 +241,93 @@ void ClientRoom::_protocol15Answer(CommunicatorMessage connectionResponse)
...
+    }
+    pseudo = connection.getPseudo();
+    password = connection.getPassword();
+    startLobbyLoop(pseudo, password);
+}
+
+void ClientRoom::startGame()
+{
+    std::cerr << "If you want to play in Solo Mod, please refer S. Otherwise if you want to play in multiplayer use M "
+                 "and be shure that a server is runing : ";
```
Comment (31Nathan) [inline Client/ClientRoom.cpp:290]: Sure and Running

File: Client/ClientRoom.cpp [hunk 4]
```diff
@@ -370,22 +482,41 @@ void ClientRoom::_initSharedResources()
...
-    _worldInstance->addSystem<AnimationSystem>();
-    _worldInstance->addSystem<NoAfkInMenu>();
-    _worldInstance->addSystem<MusicManagement>();
-    _worldInstance->addSystem<SoundManagement>();
-    _worldInstance->addSystem<RemoveChatSystem>();
+    if (isSolo) {
+        _worldInstance->addSystem<EnemiesGoRandom>();
+        _worldInstance->addSystem<EnemyShootSystem>();
+        _worldInstance->addSystem<Collide>();
+        _worldInstance->addSystem<DeathLife>();
```
Comment (31Nathan) [inline Client/ClientRoom.cpp:491]: Update Clock is in both, you should check for every system to not double them

File: Client/ClientRoom.hpp [hunk 5]
```diff
@@ -42,17 +42,33 @@ namespace client_data
...
+        /// @brief This function allows to get name and pwd of the player for the server
+        void choosePlayerInfosForServer();
+
       private:
         /// @brief Init all shared resources of the world.
         void _initSharedResources();
 
         /// @brief Init all systems of the world.
-        void _initSystems();
+        /// @param isSolo Init all systems if the player play in solo mod or in multiplayer mod
```
Comment (31Nathan) [inline Client/ClientRoom.hpp:70]: not a valid documention , isSolo boolean that tell if the game is in solo or multiplayer MODE

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.hpp [hunk 6]
```diff
@@ -22,6 +22,7 @@
 #include "R-TypeLogic/Global/Components/SizeComponent.hpp"
 #include "R-TypeLogic/Global/Components/VelocityComponent.hpp"
 #include "R-TypeLogic/Global/Components/WeightComponent.hpp"
+#include "R-TypeLogic/Global/Components/PlayerComponent.hpp"
```
Comment (31Nathan) [inline libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.hpp:25]: Why ?

File: libs/R-TypeLogic/Server/Systems/EnemyShootSystem.cpp [hunk 7]
```diff
@@ -13,7 +13,6 @@
 #include "R-TypeLogic/Global/Components/EnemyComponent.hpp"
 #include "R-TypeLogic/Global/Components/ShootingFrequencyComponent.hpp"
 #include "R-TypeLogic/Global/SharedResources/GameClock.hpp"
-
 using namespace ecs;
```
Comment (31Nathan) [inline libs/R-TypeLogic/Server/Systems/EnemyShootSystem.cpp:16]: need space

**Review by 31Nathan — CHANGES_REQUESTED** [review]

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.hpp (hunk 6 above)
Comment (TomDUVAL-MAHE) [reply libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.hpp:25]: entity.addComponent<AlliedProjectile>(ally.getComponent<Player>().playerIdentifier); in cpp l33

**Review by 31Nathan — APPROVED** [review]

File: Client/ClientRoom.cpp [hunk 8]
```diff
@@ -254,7 +359,8 @@ void ClientRoom::startLobbyLoop(const std::string &pseudo, const std::string &pa
         }
     }
     if (_state != ClientState::ENDED) {
-        _initEcsGameData();
+        _communicatorInstance.get()->sendDataToAClient(_serverEndpoint, nullptr, 0, 10);
```
Comment (osvegn) [inline Client/ClientRoom.cpp:362]: Why it appears ?

File: libs/GraphicECS/SFML/Systems/SfObjectFollowEntitySystem.hpp [hunk 9]
```diff
@@ -29,4 +29,4 @@ namespace graphicECS::SFML::Systems
     };
 } // namespace graphicECS::SFML::Systems
 
-#endif /* !SFRECTANGLEFOLLOWENTITYSYSTEM_HPP_ */
+#endif /* !_SFOBJECTFOLLOWENTITY_HPP_ */
```
Comment (osvegn) [inline libs/GraphicECS/SFML/Systems/SfObjectFollowEntitySystem.hpp:32]: Why it has been updated ?

File: libs/R-TypeLogic/Server/Systems/EnemyShootSystem.cpp [hunk 10]
```diff
@@ -25,12 +25,17 @@ void EnemyShootSystem::run(World &world)
...
+                freq.frequency = freq.baseFrequency;
+            } else {
+                createNewEnemyProjectile(world, entityPtr, "");
+                freq.frequency = freq.baseFrequency;
+            }
         }
     };
     std::for_each(joined.begin(), joined.end(), enemiesMayShoot);
-}
+ }
```
Comment (osvegn) [inline libs/R-TypeLogic/Server/Systems/EnemyShootSystem.cpp:41]: useless space

**Review by osvegn — CHANGES_REQUESTED** [review]

File: Client/ClientRoom.cpp (hunk 8 above)
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp:335]: Because it's for the lobby loop for the multiplayer mod

File: Client/ClientRoom.cpp [hunk 11]
```diff
@@ -217,6 +235,93 @@ void ClientRoom::_protocol15Answer(CommunicatorMessage connectionResponse)
...
+
+void ClientRoom::startGame()
+{
+    std::cerr << "If you want to play in Solo Mod, please refer S. Otherwise if you want to play in multiplayer use M "
+                 "and be sure that a server is running : ";
+    char choosedMod = '\0';
+
+    std::cin >> choosedMod;
+    if (choosedMod == 'S') {
+        _getClientPseudoAndPassword();
```
[the remark this answers is not in the document]
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp]: To get the name of the player

File: libs/GraphicECS/SFML/Systems/SfObjectFollowEntitySystem.hpp (hunk 9 above)
Comment (TomDUVAL-MAHE) [reply libs/GraphicECS/SFML/Systems/SfObjectFollowEntitySystem.hpp:32]: Not the good name

File: Client/ClientRoom.cpp [hunk 12]
```diff
@@ -217,6 +235,93 @@ void ClientRoom::_protocol15Answer(CommunicatorMessage connectionResponse)
...
+    std::cin >> password;
+    if (password.size() != 5) {
+        std::cerr << "Nop ! Please enter a 5 characters password.";
+        _state = ClientState::ENDED;
+        return;
+    }
+    _pseudo = pseudo;
+    _password = password;
+}
+
```
[the remark this answers is not in the document]
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp:308]: To get the name of the player

File: Client/ClientRoom.cpp [hunk 13]
```diff
@@ -217,6 +235,97 @@ void ClientRoom::_protocol15Answer(CommunicatorMessage connectionResponse)
...
+void ClientRoom::_startSoloLoop()
+{
+    std::signal(SIGINT, signalCallbackHandler);
+    _initEcsGameData(true);
+    _initSoloData();
+    _state = ClientState::IN_GAME;
+    while (_state != ClientState::ENDED && _state != ClientState::UNDEFINED) {
+        if (_state == ClientState::IN_GAME) {
+            _worldInstance.get()->runSystems(); /// WILL BE IMPROVED IN PART TWO (THREAD + CLOCK)
```
Comment (LucasTesnier) [inline Client/ClientRoom.cpp:259]: You can remove the comment ;)

File: Client/ClientRoom.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (LucasTesnier) [reply Client/ClientRoom.cpp:308]: Why did you get the player name in solo mode ? SQL base will nit be relied

File: Client/ClientRoom.cpp [hunk 14]
```diff
@@ -370,22 +480,31 @@ void ClientRoom::_initSharedResources()
     _initSpritesForEntities();
 }
 
-void ClientRoom::_initSystems()
+void ClientRoom::_initSystems(bool isSolo)
```
Comment (LucasTesnier) [inline Client/ClientRoom.cpp:483]: Why did you not use an intern variable for Solomode ?

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.cpp [hunk 15]
```diff
@@ -26,9 +26,12 @@ namespace ecs
...
             .addComponent<Damage>(damage)
             .addComponent<DamageRadius>(5)
-            .addComponent<Collidable>()
-            .addComponent<AlliedProjectile>(ally.getComponent<Networkable>().id);
-
+            .addComponent<Collidable>();
+        if (ally.contains<Networkable>()) {
+            entity.addComponent<AlliedProjectile>(ally.getComponent<Networkable>().id);
+        } else {
+            entity.addComponent<AlliedProjectile>(ally.getComponent<Player>().playerIdentifier);
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.cpp:33]: Why did you add that ?

File: Client/ClientRoom.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (31Nathan) [reply Client/ClientRoom.cpp:308]: password is useless in solo mode

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp [hunk 16]
```diff
@@ -25,17 +27,14 @@ namespace ecs
             .addComponent<DamageRadius>(damageRadius)
             .addComponent<Collidable>()
             .addComponent<Velocity>(multiplierAbscissa, multiplierOrdinate)
-            .addComponent<Enemy>(type);
-
+            .addComponent<Enemy>(type)
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp:30]: Useless in mutliplayer mod, did you envisage make a condition with an intern value of the World ?

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemyProjectile.cpp [hunk 17]
```diff
@@ -79,9 +79,12 @@ namespace ecs
...
         };
         entity.addComponent<Position>(pos.x, pos.y + 20)
-            .addComponent<Collidable>()
-            .addComponent<EnemyProjectile>(enemy.get()->getComponent<Networkable>().id);
-
+            .addComponent<Collidable>();
+        if (enemy->contains<Networkable>()) {
+            entity.addComponent<EnemyProjectile>(enemy->getComponent<Networkable>().id);
+        } else {
+            entity.addComponent<EnemyProjectile>(enemy->getComponent<Enemy>().enemyType);
```
Comment (LucasTesnier) [inline libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemyProjectile.cpp:86]: Same as Allied

File: Client/ClientRoom.hpp [hunk 18]
```diff
@@ -42,17 +42,33 @@ namespace client_data
         /// @brief Destroy the ClientRoom object
         ~ClientRoom() = default;
 
-        /// @brief Launch the lobby global loop
+        /// @brief Launch the lobby global loop for a multiplayer game
         /// @param pseudo The pseudo of the user
         /// @param password The password of the user
         void startLobbyLoop(const std::string &pseudo, const std::string &password);
 
+        /// @brief This function start the game
```
Comment (31Nathan) [inline Client/ClientRoom.hpp:50]: More details, it's start the system ? Create the entities ??

**Review by 31Nathan — CHANGES_REQUESTED** [review]

**Review by LucasTesnier — CHANGES_REQUESTED** [review]
Some question and changement

File: Client/ClientRoom.cpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp:308]: Why did you get the player name in solo mode ? SQL base will nit be relied -> For fun !

[the remark this answers is not in the document]
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp:308]: password is useless in solo mode -> yes !

File: Client/ClientRoom.cpp (hunk 14 above)
Comment (TomDUVAL-MAHE) [reply Client/ClientRoom.cpp:483]: Because we need this info into 1 function, not all the Client Room Class

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.cpp (hunk 15 above)
Comment (TomDUVAL-MAHE) [reply libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateAlliedProjectile.cpp:33]: To add the AlliedProjectileComponent with id of the player in solo mod

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp (hunk 16 above)
Comment (TomDUVAL-MAHE) [reply libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemy.cpp:30]: Enemy type is useless in multiplayer mod ???

File: libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemyProjectile.cpp (hunk 17 above)
Comment (TomDUVAL-MAHE) [reply libs/R-TypeLogic/EntityManipulation/CreateEntitiesFunctions/CreateEnemyProjectile.cpp:86]: To add the AlliedProjectileComponent with id of the player in solo mod

**Review by 31Nathan — APPROVED** [review]

**Review by LucasTesnier — APPROVED** [review]
